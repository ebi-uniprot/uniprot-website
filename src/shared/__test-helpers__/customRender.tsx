/**
 * This tries to follow what is detailed here:
 * https://testing-library.com/docs/react-testing-library/setup#custom-render
 */
import { render, RenderOptions } from '@testing-library/react';
import { Component, Dispatch, ReactElement, ReactNode } from 'react';
import { HelmetProvider } from 'react-helmet-async';
import {
  createMemoryHistory,
  LocationState,
  MemoryHistory,
  Route,
  Router,
} from 'react-router';
import { JsonValue, SetRequired } from 'type-fest';

import { MessagesState } from '../../messages/state/messagesInitialState';
import { MessagesAction } from '../../messages/state/messagesReducers';
import databaseInfo from '../../uniprotkb/utils/__tests__/__mocks__/databaseInfo';
import { getDatabaseInfoMaps } from '../../uniprotkb/utils/database';
import {
  MessagesDispatchContext,
  MessagesStateContext,
} from '../contexts/Messages';
import {
  DatabaseInfoMapsContext,
  UniProtDataVersionContext,
} from '../contexts/UniProtData';
import { JobAction } from '../workers/jobs/state/actionHandler';
import { JobsState } from '../workers/jobs/state/jobsInitialState';

type ExtraRenderOptions = {
  // For react-router
  history?: MemoryHistory<LocationState>;
  path?: string;
  // For custom user preferences (used by useLocalStorage)
  initialLocalStorage?: Record<string, JsonValue>;
  // For app context
  messagesState?: MessagesState;
  messagesDispatch?: jest.Mock<Dispatch<MessagesAction>>;
  JobsState?: JobsState;
  toolsDispatch?: jest.Mock<Dispatch<JobAction>>;
};

type WrapperProps = { children: ReactNode } & RenderOptions &
  SetRequired<
    ExtraRenderOptions,
    | 'history'
    | 'initialLocalStorage'
    | 'messagesState'
    | 'messagesDispatch'
    | 'JobsState'
    | 'toolsDispatch'
  >;

const dbInfoMaps = getDatabaseInfoMaps(databaseInfo);
const uniProtDataVersion = {
  releaseNumber: '2023_01',
  releaseDate: new Date('01-March-2023'),
};

class Wrapper extends Component<WrapperProps> {
  constructor(props: WrapperProps) {
    super(props);
    for (const [key, value] of Object.entries(props.initialLocalStorage)) {
      window.localStorage.setItem(key, JSON.stringify(value));
    }
  }

  // eslint-disable-next-line class-methods-use-this
  componentWillUnmount() {
    window.localStorage.clear();
  }

  render() {
    const { children, path, history, messagesDispatch, messagesState } =
      this.props;
    return (
      <HelmetProvider>
        <Router history={history}>
          <MessagesDispatchContext.Provider value={messagesDispatch}>
            <MessagesStateContext.Provider value={messagesState}>
              <DatabaseInfoMapsContext.Provider value={dbInfoMaps}>
                <UniProtDataVersionContext.Provider value={uniProtDataVersion}>
                  {path ? (
                    <Route path={path} render={() => children} />
                  ) : (
                    children
                  )}
                </UniProtDataVersionContext.Provider>
              </DatabaseInfoMapsContext.Provider>
            </MessagesStateContext.Provider>
          </MessagesDispatchContext.Provider>
        </Router>
      </HelmetProvider>
    );
  }
}

const customRender = (
  ui: ReactElement,
  {
    route = '',
    path,
    history = createMemoryHistory({ initialEntries: [route] }),
    initialLocalStorage = {},
    messagesState = {},
    messagesDispatch = jest.fn(),
    JobsState = {},
    toolsDispatch = jest.fn(),
    ...options
  }: RenderOptions &
    ExtraRenderOptions & {
      route?: string;
    } = {}
) => ({
  ...render(ui, {
    wrapper: (props) => (
      <Wrapper
        path={path}
        history={history}
        initialLocalStorage={initialLocalStorage}
        messagesState={messagesState}
        messagesDispatch={messagesDispatch}
        JobsState={JobsState}
        toolsDispatch={toolsDispatch}
        {...props}
      />
    ),
    ...options,
  }),
  history,
  messagesDispatch,
  toolsDispatch,
});

export default customRender;
