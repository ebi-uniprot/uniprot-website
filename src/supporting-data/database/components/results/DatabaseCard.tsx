import { useMemo } from 'react';
import { Card } from 'franklin-sites';

import RenderColumnsInCard from '../../../../shared/components/results/RenderColumnsInCard';

import { getEntryPath } from '../../../../app/config/urls';
import { getIdKeyFor } from '../../../../shared/utils/getIdKeyForNamespace';
import { mapToLinks } from '../../../../shared/components/MapTo';

import { DatabaseAPIModel } from '../../adapters/databaseConverter';
import { Namespace } from '../../../../shared/types/namespaces';
import DatabaseColumnConfiguration, {
  DatabaseColumn,
} from '../../config/DatabaseColumnConfiguration';

import renderColumnsInCardStyles from '../../../../shared/components/results/styles/render-columns-in-card.module.scss';

const category = DatabaseColumnConfiguration.get(DatabaseColumn.category);

const getIdKey = getIdKeyFor(Namespace.database);

type Props = {
  data: DatabaseAPIModel;
  selected?: boolean;
  handleEntrySelection?: (rowId: string) => void;
};

const DatabaseCard = ({ data, selected, handleEntrySelection }: Props) => {
  const id = getIdKey(data);

  const links = useMemo(
    () => mapToLinks(Namespace.database, data.abbrev, data.statistics),
    [data.statistics, data.abbrev]
  );

  return (
    <Card
      header={
        <>
          {handleEntrySelection && (
            <input
              type="checkbox"
              checked={selected}
              onChange={() => handleEntrySelection(id)}
              data-testid="up-card-checkbox"
            />
          )}
          <h2 className="tiny">{data.abbrev}</h2>
        </>
      }
      headerSeparator={false}
      to={getEntryPath(Namespace.database, id)}
      links={links}
    >
      <div className={renderColumnsInCardStyles['result-card__info-container']}>
        {data.name}
      </div>
      <RenderColumnsInCard renderers={category} data={data} />
    </Card>
  );
};

export default DatabaseCard;
