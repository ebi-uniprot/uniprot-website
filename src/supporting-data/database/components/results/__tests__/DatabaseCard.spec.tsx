import { fireEvent, screen, within } from '@testing-library/react';

import customRender from '../../../../../shared/__test-helpers__/customRender';
import databaseData from '../../../__mocks__/databaseModelData';
import DatabaseCard from '../DatabaseCard';

describe('DatabaseCard tests', () => {
  it('should render the card component', () => {
    const { asFragment } = customRender(
      <DatabaseCard data={databaseData[0]} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should allow card navigation', () => {
    const { history } = customRender(<DatabaseCard data={databaseData[0]} />);
    fireEvent.click(within(screen.getByRole('heading')).getByRole('link'));
    expect(history.location.pathname).toMatch('/database/DB-0022');
  });
});
