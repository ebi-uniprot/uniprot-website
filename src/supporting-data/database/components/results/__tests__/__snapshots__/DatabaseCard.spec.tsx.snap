// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatabaseCard tests should render the card component 1`] = `
<DocumentFragment>
  <section
    class="card"
  >
    <div
      class="card__container"
    >
      <div
        class="card__header"
      >
        <span
          class="checkbox-cell"
        >
          <input
            data-id="DB-0022"
            id="0"
            type="checkbox"
          />
          <label
            aria-label="DB-0022"
            for="0"
            title="click to select, shift+click for multiple selection"
          />
        </span>
        <h2
          class="small"
        >
          <a
            href="/database/DB-0022"
          >
            EMBL
          </a>
        </h2>
      </div>
      <div
        class="card__content"
      >
        <div
          class="result-card__info-container"
        >
          EMBL nucleotide sequence database
        </div>
        <div
          class="result-card__info-container"
        >
          <span
            class="result-card__info-bit"
          >
            <strong>
              Category: 
            </strong>
            <a
              href="/database?facets=category_exact:Sequence databases&query=*"
            >
              Sequence databases
            </a>
          </span>
        </div>
      </div>
    </div>
    <div
      class="card__actions"
    >
      <a
        class="card-action"
        href="/uniprotkb?query=(database:EMBL)"
      >
        215,511,964 UniProtKB entries
      </a>
      <a
        class="card-action"
        href="/uniprotkb?facets=reviewed:true&query=(database:EMBL)"
      >
        553,525 reviewed UniProtKB entries
      </a>
      <a
        class="card-action"
        href="/uniprotkb?facets=reviewed:false&query=(database:EMBL)"
      >
        214,958,439 unreviewed UniProtKB entries
      </a>
    </div>
  </section>
</DocumentFragment>
`;
